### PR TITLE
[Model] FalconMamba Support

### LINKED ISSUE #7478: [Bug]:  Support Falcon Mamba
### Your current environment

Does VLLM support  Falcon Mamba models? if not, when it will be supported 

### 🐛 Describe the bug

Does VLLM support  Falcon Mamba models? if not, when it will be supported

### PR DESCRIPTION
**Pull request by dhiaEddineRhaiem**
This PR adds the support for FalconMamba in VLLM.
In summary , FalconMamba is  the same Mamba class without the feed forward block and with the B_C_dt_rms LayerNorm.
Note that FalconMamba is now supported in transformers>=4.45.
I tried with this model https://huggingface.co/tiiuae/falcon-mamba-7b-instruct/ and it worked well locally.
Fixes: https://github.com/vllm-project/vllm/issues/7478

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/model_executor/models/falcon_mamba.py [hunk 1]
```diff
@@ -0,0 +1,482 @@
...
+        return next_tokens
+
+    def load_weights(self, weights: Iterable[Tuple[str, torch.Tensor]]):
+        stacked_params_mapping = [
+            # (param_name, shard_name, shard_id)
+            ("qkv_proj", "q_proj", "q"),
+            ("qkv_proj", "k_proj", "k"),
+            ("qkv_proj", "v_proj", "v"),
+            ("gate_up_proj", "gate_proj", 0),
+            ("gate_up_proj", "up_proj", 1),
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/falcon_mamba.py:449]: There's no stacked params at all, right? So I think all of this can be deleted

File: vllm/model_executor/models/falcon_mamba.py [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
+
+
+@dataclass
+class FalconMambaCacheParams:
+    is_prompt: bool = False
+    conv_state: torch.Tensor = torch.Tensor()
+    ssm_state: torch.Tensor = torch.Tensor()
+
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/falcon_mamba.py:46]: Looks like this isn't used anywhere, could you remove it? (Needs to be cleaned up in mamba.py too -- I can send a PR to do that)

File: vllm/model_executor/models/falcon_mamba.py [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+            "q_proj",
+            "k_proj",
+            "v_proj",
+        ],
+    }
+
+    # LoRA specific attributes
+    supported_lora_modules = [
+        "qkv_proj",
+        "o_proj",
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/falcon_mamba.py:335]: The references to `qkv_proj` and `o_proj` should be removed since they aren't relevant to this model

File: vllm/model_executor/models/falcon_mamba.py [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+        params_dict = dict(self.named_parameters())
+        for name, loaded_weight in weights:
+            if "rotary_emb.inv_freq" in name:
+                continue
+
+            if "A_log" in name:
+                name = name.replace("A_log", "A")
+
+            if ".self_attn." in name:
+                name = name.replace(".self_attn", "")
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/falcon_mamba.py:461]: This can be deleted as well, no?

File: vllm/model_executor/models/falcon_mamba.py [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+                    continue
+                name = name.replace(weight_name, param_name)
+                # Skip loading extra bias for GPTQ models.
+                if name.endswith(".bias") and name not in params_dict:
+                    continue
+                param = params_dict[name]
+                weight_loader = param.weight_loader
+                weight_loader(param, loaded_weight, shard_id)
+                break
+            else:
```
Comment (tlrmchlsmth) [inline vllm/model_executor/models/falcon_mamba.py:474]: And then if there are no stacked params, I  think this branch of the if statement can be removed.

**Review by tlrmchlsmth — COMMENTED** [review]
Overall looks good, thanks for the contribution!

I left a few inline comments. Biggest thing for me is to understand what's happening with the layernorm weight to see if this is the right approach.

In a subsequent PR, we should definitely factor out the mixer layer (cc @mzusman if you still plan to tackle it, or I can do it as well)

File: vllm/model_executor/layers/layernorm.py [hunk 6]
```diff
@@ -14,20 +14,24 @@ class RMSNorm(CustomOp):
...
-        self,
-        hidden_size: int,
-        eps: float = 1e-6,
-        var_hidden_size: Optional[int] = None,
-    ) -> None:
+    def __init__(self,
+                 hidden_size: int,
+                 eps: float = 1e-6,
+                 var_hidden_size: Optional[int] = None,
+                 is_learnable: bool = True) -> None:
```
[the remark this answers is not in the document]
Comment (dhiaEddineRhaiem) [reply vllm/model_executor/layers/layernorm.py:21]: 1. The idea is to register weights as parameters when they are learnable and register them as buffers whenever they are not so that they will not be included in the state_dict of the model. 
the same logic is applied here https://pytorch.org/docs/stable/_modules/torch/nn/modules/normalization.html#RMSNorm (pytorch implementation of RMSNorm)

2.Yes , you are right.

File: vllm/model_executor/models/falcon_mamba.py [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+    def __init__(self,
+                 config: FalconMambaConfig,
+                 layer_idx: int,
+                 cache_config: Optional[CacheConfig] = None,
+                 quant_config: Optional[QuantizationConfig] = None) -> None:
+        super().__init__()
+        self.layer_idx = layer_idx
+        self.config = config
+        self.mixer = FalconMambaMixer(config, layer_idx)
+
```
[the remark this answers is not in the document]
Comment (dhiaEddineRhaiem) [reply vllm/model_executor/models/falcon_mamba.py:228]: Yes, there is no feed forward section in Falcon Mamba , it is a sequence of consecutive mamba1 blocks with extra RMSNorm on input dependant parameters.

File: docs/source/models/supported_models.rst [hunk 8]
```diff
@@ -156,7 +161,7 @@ Text Generation
     - Mamba
     - :code:`state-spaces/mamba-130m-hf`, :code:`state-spaces/mamba-790m-hf`, :code:`state-spaces/mamba-2.8b-hf`, etc.
     - ✅︎
-    -
+    -
```
Comment (tlrmchlsmth) [inline docs/source/models/supported_models.rst:164]: nit: spurious whitespace

File: vllm/model_executor/layers/layernorm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (tlrmchlsmth) [reply vllm/model_executor/layers/layernorm.py:21]: Thanks for the explanation -- I think it would be better to handle this in `FalconMambaForCausalLM.load_weights`, since it's a special case that only applies to FalconMamba currently.

In load_weights, could you add a condition to check if dt_layernorm, b_layernorm, or c_layernorm is in the name? If this is the case, we can set the weight loader to a function that explicitly sets all of the elements to 1.0, which will make things explicitly clear.

**Review by tlrmchlsmth — COMMENTED** [review]
What do you think about modifying `mamba.py` to support FalconMamba instead of adding it in its own file? Seems like you would only need to handle the additional layernorms in a couple of spots

File: vllm/model_executor/layers/layernorm.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dhiaEddineRhaiem) [reply vllm/model_executor/layers/layernorm.py]: Thanks for the review

I managed to integrate FalconMamba inside mamba.py.

for rmsnorm , i reveretd the changes , but i think there is no need to handle dt_layernorm, b_layernorm, or c_layernorm inside load_weights since they have been initialised as nn.parameters(torch.ones(hidden_size)) inside RMSNorm initial implementation which is compatible with FalconMamba dt,b,c rmsnorms.

File: vllm/model_executor/layers/layernorm.py [hunk 9]
```diff
@@ -27,7 +27,6 @@ def __init__(
         self.variance_epsilon = eps
         self.variance_size_override = (None if var_hidden_size == hidden_size
                                        else var_hidden_size)
-
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/layernorm.py:30]: nit: best not to introduce whitespace-only changes to files

**Review by tlrmchlsmth — APPROVED** [review]
Looks good. Thank you for the contribution!
